### PR TITLE
Fix phpstan errors

### PR DESCRIPTION
**Pull request by BenMorel**
|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | no
| Fixed issues | none

#### Summary

A method declared with a `string` return type attempts to return `null`. This is not caught by tests as this method is never called because `SqlitePlatform` does not support exceptions.

This now throws a `LogicException` instead, as this is not expected to be executed.

#### Update

Following the discussion below, this now fixes many more phpstan errors.

### CODE REVIEW

File: tests/Doctrine/Tests/DBAL/Platforms/SqlitePlatformTest.php
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Platforms/SqlitePlatformTest.php]: > This is not caught by tests as this method is never called because `SqlitePlatform` does not support exceptions.

I'm curious why this is not flagged by PHPStan. I'm not a fan of fixing issues like this one by one. There may be more issues like this including the production code.

As for the code change, I'd rather fail the test explicitly with `$this->fail()`.

Comment (BenMorel) [reply tests/Doctrine/Tests/DBAL/Platforms/SqlitePlatformTest.php]: > I'm curious why this is not flagged by PHPStan.

I've seen your work related to phpstan earlier. Did you run it on `tests`, not just `lib`?

> As for the code change, I'd rather fail the test explicitly with $this->fail().

I'll happily change it if you think it's better, but be advised that it makes it harder to understand for static analysis, as the method does not return or throw anything. The throw is hidden in `fail()`.

Comment (morozov) [reply tests/Doctrine/Tests/DBAL/Platforms/SqlitePlatformTest.php]: > I've seen your work related to phpstan earlier. Did you run it on `tests`, not just `lib`?

It should run on `lib` and `tests`:

https://github.com/doctrine/dbal/blob/644531f360e48d4f222ccfd726a07aed60200b3e/phpstan.neon.dist#L4-L5

> […] it makes it harder to understand for static analysis, as the method does not return or throw anything.

AFAIK the PHPStan extension for PHPUnit is aware of methods like that.

> I'll happily change […]

I'd like to make sure we understand the root cause of why it's not detected before making any change.

Comment (BenMorel) [reply tests/Doctrine/Tests/DBAL/Platforms/SqlitePlatformTest.php]: > I'd like to make sure we understand the root cause of why it's not detected before making any change.

I'll leave it to you then, I'm not familiar enough with phpstan!

Comment (BenMorel) [reply tests/Doctrine/Tests/DBAL/Platforms/SqlitePlatformTest.php]: > As for the code change, I'd rather fail the test explicitly with `$this->fail()`.

Done, FWIW.

Comment (morozov) [reply tests/Doctrine/Tests/DBAL/Platforms/SqlitePlatformTest.php]: Scrutinizer sucks.

> […] the function will implicitly return `null` […]

It is not true. The return will never happen.

Comment (SenseException) [reply tests/Doctrine/Tests/DBAL/Platforms/SqlitePlatformTest.php]: > I'd like to make sure we understand the root cause of why it's not detected before making any change.

@morozov Are you going for the root cause in this PR or will you create a separate one?

Comment (morozov) [reply tests/Doctrine/Tests/DBAL/Platforms/SqlitePlatformTest.php]: I'd like to address the root cause. Apparently, PHPStan ignores this issue because of:
https://github.com/doctrine/dbal/blob/f9b3e7908f548a96d4ec62a30c8cb9fed7f95c7f/phpstan.neon.dist#L38-L39

Instead of being applied to a specific class method, it's applied to the entire codebase. Removing this line makes PHPStan find the issue as well as 43 others:
```
$ phpstan a tests/Doctrine/Tests/DBAL/Platforms/SqlitePlatformTest.php

 ------ ------------------------------------------------------------------------------------- 
  Line   SqlitePlatformTest.php                                                               
 ------ ------------------------------------------------------------------------------------- 
  274    Method Doctrine\Tests\DBAL\Platforms\SqlitePlatformTest::getGenerateForeignKeySql()  
         should return string but returns null.                                               
 ------ ------------------------------------------------------------------------------------- 
```
@BenMorel please try to rework the PHPStan configuration so that it's only applied to where it's supposed to and try to fix the rest of the issues.

Comment (morozov) [reply tests/Doctrine/Tests/DBAL/Platforms/SqlitePlatformTest.php]: The problem is probably because the `|` in the `string|null` type is not escaped, so it's interpreted by the regex engine instead of being used as a literal. We should escape all occurrences of `|` where it's part of the expected error message like it's done it:
https://github.com/doctrine/dbal/blob/f9b3e7908f548a96d4ec62a30c8cb9fed7f95c7f/phpstan.neon.dist#L16-L17

Comment (BenMorel) [reply tests/Doctrine/Tests/DBAL/Platforms/SqlitePlatformTest.php]: First of all, when I run `phpstan analyze` locally, I get several times this warning:

> Access to undefined constant PDO::PGSQL_ATTR_DISABLE_PREPARES

How do I get rid of it? Should I report a bug with phpstan?

Comment (BenMorel) [reply tests/Doctrine/Tests/DBAL/Platforms/SqlitePlatformTest.php]: > The problem is probably because the | in the string|null type is not escaped, so it's interpreted by the regex engine instead of being used as a literal.

Good catch! Fixed this in 8f286d31c5b48d58ae71d47a02e3d4fafd279ef5, and now phpstan reports 42 errors. Working on fixing them.

File: lib/Doctrine/DBAL/Connection.php
Comment (BenMorel) [inline lib/Doctrine/DBAL/Connection.php]: I'm not sure what your policy is in this case: `assert()`? `throw new LogicException()?`

There are several other places, as you'll see in the diff, where we know that the variable will not be null, but phpstan cannot dig deep enough to figure it out by itself.

File: lib/Doctrine/DBAL/Portability/Statement.php
Comment (BenMorel) [inline lib/Doctrine/DBAL/Portability/Statement.php]: This variable [is expected to be nullable](https://github.com/doctrine/dbal/blob/b6e3aa3555e81d3168b4ba96db62ff3f31b93690/lib/Doctrine/DBAL/Portability/Statement.php#L121), however `array_change_key_case()` below expects it to be `int`, never `null`. Hence [another `assert()` below](https://github.com/doctrine/dbal/pull/3801/files#diff-87c7f450dc8a06beead2da33fd34b210R179).

File: lib/Doctrine/DBAL/Schema/MySqlSchemaManager.php
Comment (BenMorel) [inline lib/Doctrine/DBAL/Schema/MySqlSchemaManager.php]: Removed this one as the function body is the same as the parent class.

**Comment by BenMorel** [comment]
All phpstan errors fixed now! @morozov, waiting for your feedback on assertions, in particular.

File: lib/Doctrine/DBAL/Portability/Statement.php
Comment (morozov) [reply lib/Doctrine/DBAL/Portability/Statement.php]: This also seems to be the case for 

https://github.com/doctrine/dbal/blob/88f7feae9a74a1f804956470e75284804944d240/lib/Doctrine/DBAL/Portability/Connection.php#L38-L39

File: lib/Doctrine/DBAL/Connection.php
Comment (morozov) [reply lib/Doctrine/DBAL/Connection.php]: The assertions should be added only if the issue reported by PHPStan is in fact impossible but this is not obvious from the type system. On the other hand, if an exception is added, there should be a test case that covers it being thrown.

In this case, the assertion looks legit. While the API could use some improvement. There should be a method that always _returns_ a connection, not sets it on the object (out of scope).

File: lib/Doctrine/DBAL/Platforms/SqlitePlatform.php
Comment (morozov) [inline lib/Doctrine/DBAL/Platforms/SqlitePlatform.php]: Instead of adding an assertion against the object state, if we're in a private method, would it make sense to pass the value explicitly and enforce its type in the method signature?

File: lib/Doctrine/DBAL/Platforms/SqlitePlatform.php
Comment (morozov) [inline lib/Doctrine/DBAL/Platforms/SqlitePlatform.php]: Please see if replacing the property access with a method argument would make the API more explicit and eliminate the need for the assertion.

File: lib/Doctrine/DBAL/Portability/Statement.php
Comment (morozov) [inline lib/Doctrine/DBAL/Portability/Statement.php]: Here we could use the same approach as with private methods above, however, this one is protected. We should see what prevents this class from being finalized (https://github.com/doctrine/dbal/pull/3590).

File: lib/Doctrine/DBAL/Query/QueryBuilder.php
Comment (morozov) [inline lib/Doctrine/DBAL/Query/QueryBuilder.php]: :+1:

**Review by morozov — COMMENTED** [review]
@BenMorel I reviewed this partially. Please see my recommendations on using assertions and see if the rest of the assertions follow them.

File: lib/Doctrine/DBAL/Portability/Statement.php
Comment (BenMorel) [reply lib/Doctrine/DBAL/Portability/Statement.php]: Indeed, it's null by default (not set in a constructor). And the getter reflects this. Fixed in ff9831475dca77af2e150efd8ad473ad4922b7ce.

File: lib/Doctrine/DBAL/Platforms/SqlitePlatform.php
Comment (BenMorel) [reply lib/Doctrine/DBAL/Platforms/SqlitePlatform.php]: I think this is doable, as the calling methods seem to have checks for `$fromTable` being not null, for example:

https://github.com/doctrine/dbal/blob/69cd77e007466b8dd9b671c106c419fee5e2d877/lib/Doctrine/DBAL/Platforms/SqlitePlatform.php#L825-L828

This means passing as a parameter a property of the object itself, though, and I feel like it's a smell. Are you sure you want to try this route?

File: lib/Doctrine/DBAL/Platforms/SqlitePlatform.php
Comment (BenMorel) [reply lib/Doctrine/DBAL/Platforms/SqlitePlatform.php]: Please see above. Doable at first glance. But is it a good practice?

File: lib/Doctrine/DBAL/Platforms/SqlitePlatform.php
Comment (morozov) [reply lib/Doctrine/DBAL/Platforms/SqlitePlatform.php]: It is a smell (probably, the method doesn't belong to the class or should be made static) but assertions are also smell.

Comment (BenMorel) [reply lib/Doctrine/DBAL/Platforms/SqlitePlatform.php]: Alright then! Done in 4815eb8b33777b1dec2666247146c43643d572c9. Not a big deal, after all.

File: tests/Doctrine/Tests/DBAL/Platforms/SqlitePlatformTest.php
Comment (morozov) [reply tests/Doctrine/Tests/DBAL/Platforms/SqlitePlatformTest.php]: > Should I report a bug with phpstan?

Just for a future reference: this is a current limitation of PHPStan and the design flaw of PDO. Some PDO extensions define constants on the already existing `PDO` class. If `pdo` is loaded but `pdo_pgsql` isn't, PHPStan cannot add the missing constants from the `pdo_pgsql` stubs because it uses runtime reflection instead of static.

File: lib/Doctrine/DBAL/Portability/Statement.php
Comment (BenMorel) [reply lib/Doctrine/DBAL/Portability/Statement.php]: What should we do here then? There are only 4 `assert()` left in this PR; do you want to find a solution to get rid of each of them individually, or is it OK if they stay?

Comment (morozov) [reply lib/Doctrine/DBAL/Portability/Statement.php]: Let's leave those assertions for now. Please rebase and squash.

**Comment by morozov** [comment]
@BenMorel once this is done, could you please also take a look at https://github.com/doctrine/dbal/blob/f9b3e7908f548a96d4ec62a30c8cb9fed7f95c7f/phpstan.neon.dist#L35-L36

If the `|` is properly escaped, PHPStan produces 14 more errors that need to be fixed or properly suppressed.

File: lib/Doctrine/DBAL/Portability/Statement.php
Comment (BenMorel) [reply lib/Doctrine/DBAL/Portability/Statement.php]: Rebased & squashed!

**Comment by BenMorel** [comment]
> @BenMorel once this is done, could you please also take a look at (...)

I'll try to have a look tomorrow!

**Review by morozov — APPROVED** [review]

**Comment by morozov** [comment]
Thanks @BenMorel. Great work :+1:

**Pull request merged**
